<html>
<head>
  <title>VEP summary</title>
  <script type="text/javascript" src="http://www.google.com/jsapi"></script>
  <script type="text/javascript">
    google.load('visualization', '1', {packages: ['corechart','table']});
  </script>
  <script type="text/javascript">
    
    function init() {
      // charts
      var var_class_pie = drawPie('var_class_pie', 'Variant classes', google.visualization.arrayToDataTable([['Variant class','Count'],['insertion',516],['deletion',517],['SNV',7762]]), null);
var var_class_table = drawTable('var_class_table', 'Variant classes', google.visualization.arrayToDataTable([['Variant class','Count'],['insertion',516],['deletion',517],['SNV',7762]]));

          google.visualization.events.addListener(var_class_pie, 'select', function() {
            var_class_table.setSelection(var_class_pie.getSelection());
          });
          google.visualization.events.addListener(var_class_table, 'select', function() {
            var_class_pie.setSelection(var_class_table.getSelection());
          });
        var var_cons_pie = drawPie('var_cons_pie', 'Consequences (most severe)', google.visualization.arrayToDataTable([['Consequence type','Count'],['splice_acceptor_variant',3],['splice_donor_variant',6],['stop_gained',16],['frameshift_variant',24],['stop_lost',3],['start_lost',6],['inframe_insertion',17],['inframe_deletion',18],['missense_variant',657],['protein_altering_variant',2],['splice_donor_5th_base_variant',4],['splice_region_variant',122],['splice_donor_region_variant',17],['splice_polypyrimidine_tract_variant',107],['synonymous_variant',669],['5_prime_UTR_variant',346],['3_prime_UTR_variant',553],['non_coding_transcript_exon_variant',948],['intron_variant',4857],['upstream_gene_variant',209],['downstream_gene_variant',131],['intergenic_variant',80]]), {slices: [{color: "#ff7f50"}, {color: "#ff7f50"}, {color: "#ff0000"}, {color: "#ff69b4"}, {color: "#ff0000"}, {}, {color: "#ff69b4"}, {color: "#ff69b4"}, {color: "#ffd700"}, {}, {}, {color: "#ff7f50"}, {}, {}, {color: "#76ee00"}, {}, {}, {}, {color: "#02599c"}, {color: "#a2b5cd"}, {color: "#a2b5cd"}, {color: "gray"}]});
var var_cons_table = drawTable('var_cons_table', 'Consequences (most severe)', google.visualization.arrayToDataTable([['Consequence type','Count'],['splice_acceptor_variant',3],['splice_donor_variant',6],['stop_gained',16],['frameshift_variant',24],['stop_lost',3],['start_lost',6],['inframe_insertion',17],['inframe_deletion',18],['missense_variant',657],['protein_altering_variant',2],['splice_donor_5th_base_variant',4],['splice_region_variant',122],['splice_donor_region_variant',17],['splice_polypyrimidine_tract_variant',107],['synonymous_variant',669],['5_prime_UTR_variant',346],['3_prime_UTR_variant',553],['non_coding_transcript_exon_variant',948],['intron_variant',4857],['upstream_gene_variant',209],['downstream_gene_variant',131],['intergenic_variant',80]]));

          google.visualization.events.addListener(var_cons_pie, 'select', function() {
            var_cons_table.setSelection(var_cons_pie.getSelection());
          });
          google.visualization.events.addListener(var_cons_table, 'select', function() {
            var_cons_pie.setSelection(var_cons_table.getSelection());
          });
        var consequences_pie = drawPie('consequences_pie', 'Consequences (all)', google.visualization.arrayToDataTable([['Consequence type','Count'],['splice_acceptor_variant',4],['splice_donor_variant',23],['stop_gained',44],['frameshift_variant',83],['stop_lost',3],['start_lost',18],['inframe_insertion',37],['inframe_deletion',89],['missense_variant',3034],['protein_altering_variant',3],['splice_donor_5th_base_variant',22],['splice_region_variant',1006],['splice_donor_region_variant',76],['splice_polypyrimidine_tract_variant',1341],['start_retained_variant',10],['stop_retained_variant',1],['synonymous_variant',3626],['coding_sequence_variant',8],['5_prime_UTR_variant',1479],['3_prime_UTR_variant',2833],['non_coding_transcript_exon_variant',3236],['intron_variant',50219],['NMD_transcript_variant',2735],['non_coding_transcript_variant',8336],['upstream_gene_variant',10537],['downstream_gene_variant',13458],['intergenic_variant',89]]), {slices: [{color: "#ff7f50"}, {color: "#ff7f50"}, {color: "#ff0000"}, {color: "#ff69b4"}, {color: "#ff0000"}, {}, {color: "#ff69b4"}, {color: "#ff69b4"}, {color: "#ffd700"}, {}, {}, {color: "#ff7f50"}, {}, {}, {}, {color: "#76ee00"}, {color: "#76ee00"}, {color: "#458b00"}, {}, {}, {}, {color: "#02599c"}, {}, {}, {color: "#a2b5cd"}, {color: "#a2b5cd"}, {color: "gray"}]});
var consequences_table = drawTable('consequences_table', 'Consequences (all)', google.visualization.arrayToDataTable([['Consequence type','Count'],['splice_acceptor_variant',4],['splice_donor_variant',23],['stop_gained',44],['frameshift_variant',83],['stop_lost',3],['start_lost',18],['inframe_insertion',37],['inframe_deletion',89],['missense_variant',3034],['protein_altering_variant',3],['splice_donor_5th_base_variant',22],['splice_region_variant',1006],['splice_donor_region_variant',76],['splice_polypyrimidine_tract_variant',1341],['start_retained_variant',10],['stop_retained_variant',1],['synonymous_variant',3626],['coding_sequence_variant',8],['5_prime_UTR_variant',1479],['3_prime_UTR_variant',2833],['non_coding_transcript_exon_variant',3236],['intron_variant',50219],['NMD_transcript_variant',2735],['non_coding_transcript_variant',8336],['upstream_gene_variant',10537],['downstream_gene_variant',13458],['intergenic_variant',89]]));

          google.visualization.events.addListener(consequences_pie, 'select', function() {
            consequences_table.setSelection(consequences_pie.getSelection());
          });
          google.visualization.events.addListener(consequences_table, 'select', function() {
            consequences_pie.setSelection(consequences_table.getSelection());
          });
        var coding_pie = drawPie('coding_pie', 'Coding consequences', google.visualization.arrayToDataTable([['Consequence type','Count'],['stop_gained',44],['frameshift_variant',83],['stop_lost',3],['start_lost',18],['inframe_insertion',37],['inframe_deletion',89],['missense_variant',3034],['protein_altering_variant',3],['start_retained_variant',10],['stop_retained_variant',1],['synonymous_variant',3626],['coding_sequence_variant',8]]), {slices: [{color: "#ff0000"}, {color: "#ff69b4"}, {color: "#ff0000"}, {}, {color: "#ff69b4"}, {color: "#ff69b4"}, {color: "#ffd700"}, {}, {}, {color: "#76ee00"}, {color: "#76ee00"}, {color: "#458b00"}]});
var coding_table = drawTable('coding_table', 'Coding consequences', google.visualization.arrayToDataTable([['Consequence type','Count'],['stop_gained',44],['frameshift_variant',83],['stop_lost',3],['start_lost',18],['inframe_insertion',37],['inframe_deletion',89],['missense_variant',3034],['protein_altering_variant',3],['start_retained_variant',10],['stop_retained_variant',1],['synonymous_variant',3626],['coding_sequence_variant',8]]));

          google.visualization.events.addListener(coding_pie, 'select', function() {
            coding_table.setSelection(coding_pie.getSelection());
          });
          google.visualization.events.addListener(coding_table, 'select', function() {
            coding_pie.setSelection(coding_table.getSelection());
          });
        var chr_bar = drawBar('chr_bar', 'Variants by chromosome', google.visualization.arrayToDataTable([['Chromosome','Count'],['chr22',5251],['chrX',3340],['chrY',204]]), {legend: {position: "none"}});
var chr_table = drawTable('chr_table', 'Variants by chromosome', google.visualization.arrayToDataTable([['Chromosome','Count'],['chr22',5251],['chrX',3340],['chrY',204]]));

          google.visualization.events.addListener(chr_bar, 'select', function() {
            chr_table.setSelection(chr_bar.getSelection());
          });
          google.visualization.events.addListener(chr_table, 'select', function() {
            chr_bar.setSelection(chr_table.getSelection());
          });
        var chr_chr22_area = drawArea('chr_chr22_area', 'Distribution of variants on chromosome chr22', google.visualization.arrayToDataTable([['Position (mb)','Count'],['0',0],['1',0],['2',0],['3',0],['4',0],['5',0],['6',0],['7',0],['8',0],['9',0],['10',0],['11',0],['12',0],['13',0],['14',0],['15',0],['16',122],['17',147],['18',256],['19',200],['20',268],['21',343],['22',141],['23',136],['24',228],['25',148],['26',155],['27',9],['28',18],['29',157],['30',176],['31',188],['32',98],['33',11],['34',6],['35',45],['36',175],['37',212],['38',208],['39',168],['40',98],['41',118],['42',219],['43',174],['44',145],['45',195],['46',73],['47',78],['48',5],['49',12],['50',442],['51',77]]), {hAxis: {title: "Position (mb)", textStyle: {fontSize: 8}}, legend: {position: "none"}});
var chr_chrX_area = drawArea('chr_chrX_area', 'Distribution of variants on chromosome chrX', google.visualization.arrayToDataTable([['Position (mb)','Count'],['0',60],['1',74],['2',165],['3',38],['4',0],['5',10],['6',12],['7',16],['8',32],['9',27],['10',26],['11',10],['12',24],['13',34],['14',10],['15',56],['16',22],['17',10],['18',23],['19',11],['20',2],['21',17],['22',9],['23',55],['24',28],['25',3],['26',4],['27',11],['28',4],['29',1],['30',27],['31',25],['32',24],['33',4],['34',7],['35',20],['36',5],['37',8],['38',15],['39',8],['40',16],['41',46],['42',0],['43',19],['44',15],['45',13],['46',21],['47',53],['48',97],['49',59],['50',19],['51',7],['52',40],['53',38],['54',36],['55',23],['56',10],['57',36],['58',0],['59',0],['60',0],['61',13],['62',0],['63',3],['64',1],['65',12],['66',11],['67',11],['68',10],['69',72],['70',60],['71',53],['72',9],['73',13],['74',18],['75',2],['76',6],['77',13],['78',16],['79',18],['80',6],['81',0],['82',3],['83',8],['84',28],['85',6],['86',14],['87',1],['88',2],['89',13],['90',0],['91',11],['92',6],['93',0],['94',0],['95',7],['96',13],['97',0],['98',0],['99',25],['100',77],['101',39],['102',18],['103',21],['104',6],['105',7],['106',25],['107',62],['108',9],['109',2],['110',2],['111',2],['112',1],['113',1],['114',13],['115',8],['116',4],['117',58],['118',62],['119',42],['120',37],['121',0],['122',14],['123',27],['124',0],['125',9],['126',0],['127',1],['128',44],['129',21],['130',16],['131',2],['132',8],['133',6],['134',49],['135',54],['136',4],['137',24],['138',17],['139',11],['140',37],['141',2],['142',17],['143',1],['144',13],['145',3],['146',3],['147',16],['148',42],['149',21],['150',26],['151',35],['152',98],['153',179],['154',31],['155',44]]), {hAxis: {title: "Position (mb)", textStyle: {fontSize: 8}}, legend: {position: "none"}});
var chr_chrY_area = drawArea('chr_chrY_area', 'Distribution of variants on chromosome chrY', google.visualization.arrayToDataTable([['Position (mb)','Count'],['0',61],['1',74],['2',24],['3',0],['4',0],['5',0],['6',0],['7',0],['8',0],['9',0],['10',0],['11',0],['12',0],['13',0],['14',0],['15',0],['16',0],['17',0],['18',0],['19',0],['20',0],['21',0],['22',0],['23',0],['24',0],['25',0],['26',0],['27',0],['28',0],['29',0],['30',0],['31',0],['32',0],['33',0],['34',0],['35',0],['36',0],['37',0],['38',0],['39',0],['40',0],['41',0],['42',0],['43',0],['44',0],['45',0],['46',0],['47',0],['48',0],['49',0],['50',0],['51',0],['52',0],['53',0],['54',0],['55',0],['56',0],['57',0],['58',0],['59',45]]), {hAxis: {title: "Position (mb)", textStyle: {fontSize: 8}}, legend: {position: "none"}});
var protein_bar = drawBar('protein_bar', 'Position in protein', google.visualization.arrayToDataTable([['Position in protein (percentile)','Count'],['00-10%',697],['10-20%',610],['20-30%',575],['30-40%',698],['40-50%',686],['50-60%',695],['60-70%',767],['70-80%',680],['80-90%',879],['90-100%',640]]), {hAxis: {title: "Position in protein (percentile)", textStyle: {fontSize: 10}}, legend: {position: "none"}});

    }
    
    function drawPie(id, title, data, options) {    
      var pie = new google.visualization.PieChart(document.getElementById(id));
      pie.draw(data, options);
      return pie;
    }
    function drawBar(id, title, data, options) {
      var bar = new google.visualization.ColumnChart(document.getElementById(id));
      bar.draw(data, options);
      return bar;
    }
    function drawTable(id, title, data) {
      var table = new google.visualization.Table(document.getElementById(id));
      table.draw(data, null);
      return table;
    }
    function drawLine(id, title, data, options) {
      var line = new google.visualization.LineChart(document.getElementById(id));
      line.draw(data, options);
      return line;
    }
    function drawArea(id, title, data, options) {
      var area = new google.visualization.AreaChart(document.getElementById(id));
      area.draw(data, options);
      return area;
    }
    google.setOnLoadCallback(init);
  </script>
  
  
  <style type="text/css">
    body {
      font-family: arial, sans-serif;
      margin: 0px;
      padding: 0px;
    }
    
    a {color: #36b;}
    a.visited {color: #006;}
    
    .stats_table {
      margin: 5px;
      table-layout: fixed;
      width: 100%;
    }

    pre {
      white-space: normal;
      word-break: keep-all;
    }

    td {
      padding: 5px;
      word-break: break-word;
    }

    td:nth-child(2) {
      width: 80%;
    }

    th.gradient {
      height: auto;
    }

    .stats_table tr:nth-child(odd) {
      background-color: #f0f0f0;
    }
    
    h3 {
      color: #666;
    }
    
    .masthead {
      background-color: white;
      color: rgb(204, 221, 255);
      height: 80px;
      width: 100%;
      padding: 0px;
    }
    
    .main {
      padding: 10px;
    }
    
    .gradient {
      background: #333366; /* Old browsers */
      background: -moz-linear-gradient(left,  #333366 0%, #ffffff 100%); /* FF3.6+ */
      background: -webkit-gradient(linear, left top, right top, color-stop(0%,#333366), color-stop(100%,#ffffff)); /* Chrome,Safari4+ */
      background: -webkit-linear-gradient(left,  #333366 0%,#ffffff 100%); /* Chrome10+,Safari5.1+ */
      background: -o-linear-gradient(left,  #333366 0%,#ffffff 100%); /* Opera 11.10+ */
      background: -ms-linear-gradient(left,  #333366 0%,#ffffff 100%); /* IE10+ */
      background: linear-gradient(to right,  #333366 0%,#ffffff 100%); /* W3C */
      filter: progid:DXImageTransform.Microsoft.gradient( startColorstr='#333366', endColorstr='#ffffff',GradientType=1 ); /* IE6-9 */
      
      padding: 0px;
      height: 80px;
      width: 700px;
    }
    
    .main_content {
      margin-left: 300px;
    }
    
    .sidemenu {
      width: 260px;
      position: fixed;
      border-style: solid;
      border-width: 2px;
      border-color: rgb(51, 51, 102);
    }
    
    .sidemenu_head {
      width: 250px;
      background-color: rgb(51, 51, 102);
      color: rgb(204, 221, 255);
      padding: 5px;
    }
    
    .sidemenu_body {
      width: 250px;
      padding: 5px;
    }
  </style>
</head>
<body>
<div id="masthead" class="masthead">
  <div style="float: left; display: inline; padding: 10px; height: 80px;">
    <a href="http://www.ensembl.org/"><img src="http://static.ensembl.org/i/e-ensembl.png" width="185" height="45" alt="Ensembl"></a>
  </div>
  
  <div style="float: right; display: inline; height: 80px; background: white; padding: 10px;">
    <a href="http://www.ensembl.org/vep"><img src="http://www.ensembl.org/img/vep_logo.png" width="111" height="50" alt="VEP"></a>
  </div>
  <div class="gradient">
  </div>
</div>
<div class="main">
<div class="sidemenu"><div class="sidemenu_head">Links</div><div class="sidemenu_body"><ul><li><a href="#masthead">Top of page</a></li><li><a href="#run_stats">VEP run statistics</a></li><li><a href="#data_version">Data version</a></li><li><a href="#gen_stats">General statistics</a></li><li><a href="#var_class">Variant classes</a></li><li><a href="#var_cons">Consequences (most severe)</a></li><li><a href="#consequences">Consequences (all)</a></li><li><a href="#coding">Coding consequences</a></li><li><a href="#chr">Variants by chromosome</a></li><li><a href="#protein">Position in protein</a></li></ul></div></div><div class='main_content'><h3 id="run_stats">VEP run statistics</h3><table class="stats_table"><tr><td>VEP version (API)</td><td> 112.0 (112)</td></tr><tr><td>Annotation sources</td><td>Cache: <kbd>/home/yangyxt/Caches/ensembl-vep/homo_sapiens_merged/112_GRCh37</kbd></td></tr><tr><td>Species</td><td>homo_sapiens</td></tr><tr><td>Command line options</td><td><pre>--assembly GRCh37 --buffer_size 10000 --cache  --canonical --dir_cache /home/yangyxt/Caches/ensembl-vep --dir_plugins /home/yangyxt/Tools/VEP_plugins --fasta /home/yangyxt/Caches/genomes/hg19/ucsc.hg19.fasta --force_overwrite --fork 5 --format vcf --gene_phenotype --hgvs --input_file /home/yangyxt/git_repos/acmg_auto/data/test_example/TEST_FAM.filtered.anno.vcf.gz --merged --numbers --offline --output_file /home/yangyxt/git_repos/acmg_auto/data/test_example/TEST_FAM.filtered.anno._K5fC_cBl_FjNo8KJGSCzgfl7lFPcsEo.vcf --plugin UTRAnnotator,file=/home/yangyxt/Caches/VEP_plugin_caches/UTRannotator/uORF_5UTR_GRCh37_PUBLIC.txt --plugin LOEUF,file=/home/yangyxt/Caches/VEP_plugin_caches/LOEUF/loeuf_dataset.tsv.gz,match_by=transcript --plugin AlphaMissense,file=/home/yangyxt/Caches/VEP_plugin_caches/AlphaMissense/AlphaMissense_hg19.tsv.gz --plugin SpliceAI,snv=/home/yangyxt/Caches/VEP_plugin_caches/SpliceAI/hg19/spliceai_scores.raw.snv.hg19.vcf.gz,indel=/home/yangyxt/Caches/VEP_plugin_caches/SpliceAI/hg19/spliceai_scores.raw.indel.hg19.vcf.gz,cutoff=0.5 --plugin PrimateAI,/home/yangyxt/Caches/VEP_plugin_caches/PrimateAI/hg19/PrimateAI_scores_v0.2_GRCh37_sorted.tsv.bgz --plugin Conservation,/home/yangyxt/Caches/VEP_plugin_caches/Conservations/gerp_conservation_scores.homo_sapiens.GRCh37.bw,MAX --plugin NMD --species homo_sapiens --stats_file /home/yangyxt/git_repos/acmg_auto/data/test_example/TEST_FAM.filtered.anno.vep.stats.html --symbol --use_given_ref --variant_class --vcf --verbose</pre></td></tr><tr><td>Start time</td><td>2024-11-12 17:04:27</td></tr><tr><td>End time</td><td>2024-11-12 17:07:54</td></tr><tr><td>Run time</td><td>207 seconds</td></tr><tr><td>Input file</td><td><pre>/home/yangyxt/git_repos/acmg_auto/data/test_example/TEST_FAM.filtered.anno.vcf.gz</pre></td></tr><tr><td>Output file</td><td><pre>/home/yangyxt/git_repos/acmg_auto/data/test_example/TEST_FAM.filtered.anno._K5fC_cBl_FjNo8KJGSCzgfl7lFPcsEo.vcf</pre></td></tr></table><h3 id="data_version">Data version</h3><table class="stats_table"><tr><td>1000genomes</td><td>phase3</td></tr><tr><td>COSMIC</td><td>98</td></tr><tr><td>ClinVar</td><td>202306</td></tr><tr><td>HGMD-PUBLIC</td><td>20204</td></tr><tr><td>assembly</td><td>GRCh37.p13</td></tr><tr><td>dbSNP</td><td>156</td></tr><tr><td>gencode</td><td>GENCODE 19</td></tr><tr><td>genebuild</td><td>2011-04</td></tr><tr><td>gnomADe</td><td>r2.1</td></tr><tr><td>polyphen</td><td>2.2.2</td></tr><tr><td>refseq</td><td>105.20220307 - GCF_000001405.25_GRCh37.p13_genomic.gff</td></tr><tr><td>regbuild</td><td>1.0</td></tr><tr><td>sift</td><td>sift5.2.2</td></tr></table><h3 id="gen_stats">General statistics</h3><table class="stats_table"><tr><td>Lines of input read</td><td>8795</td></tr><tr><td>Variants processed</td><td>8795</td></tr><tr><td>Variants filtered out</td><td>0</td></tr><tr><td>Novel / existing variants</td><td>-</td></tr><tr><td>Overlapped genes</td><td>3089</td></tr><tr><td>Overlapped transcripts</td><td>11056</td></tr><tr><td>Overlapped regulatory features</td><td>-</td></tr></table><hr/><h3 id="var_class">Variant classes</h3><div id="var_class_pie" style="width: 800px; height: 200px">&nbsp;</div><div id="var_class_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="var_cons">Consequences (most severe)</h3><div id="var_cons_pie" style="width: 800px; height: 400px">&nbsp;</div><div id="var_cons_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="consequences">Consequences (all)</h3><div id="consequences_pie" style="width: 800px; height: 400px">&nbsp;</div><div id="consequences_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="coding">Coding consequences</h3><div id="coding_pie" style="width: 800px; height: 400px">&nbsp;</div><div id="coding_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="chr">Variants by chromosome</h3><div id="chr_bar" style="width: 800px; height: 200px">&nbsp;</div><div id="chr_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="chr_chr22">Distribution of variants on chromosome chr22</h3><div id="chr_chr22_area" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="chr_chrX">Distribution of variants on chromosome chrX</h3><div id="chr_chrX_area" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="chr_chrY">Distribution of variants on chromosome chrY</h3><div id="chr_chrY_area" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="protein">Position in protein</h3><div id="protein_bar" style="width: 800px; height: 200px">&nbsp;</div></div>
</div></body>
</html>
